rust/relay/src/dual_stack_socket.rs

```diff
@@ -0,0 +1,84 @@
+use anyhow::{Context, Result};
+use local_ip_address::{local_ip, local_ipv6};
+use std::net::{IpAddr, SocketAddr, SocketAddrV4, SocketAddrV6};
+use std::pin::pin;
+use tokio::net::UdpSocket;
+
+const MAX_UDP_SIZE: usize = 65536;
+
+/// A thin abstraction over [`UdpSocket`] that always listens on IPv4 as well as IPv6.
+pub struct DualStackSocket {
+    ip4_socket: UdpSocket,
+    ip6_socket: UdpSocket,
+
+    ip4_addr: SocketAddrV4,
+    ip6_addr: SocketAddrV6,
+
+    ip4_receive_buffer: [u8; MAX_UDP_SIZE],
+    ip6_receive_buffer: [u8; MAX_UDP_SIZE],
+}
+
+impl DualStackSocket {
+    pub async fn listen_on(port: u16) -> Result<Self> {
+        let ip4_addr = local_ip().context("failed to retrieve local IPv4 address")?;
+        let ip6_addr = local_ipv6().context("failed to retrieve local IPv6 address")?;
```

Comment:
Are we planning to run this on machines with more than one interface?

The thing is, the OS only lets me bind to the address of the local NIC, not the public one. At the same time, I cannot bind to `0.0.0.0` because it is kernel dependent whether that listens on IPv6 or not.

I couldn't figure out a better way to do this, open to suggestions!

Reply:
Yeah there will most likely be multiple interfaces available in deployment env. Maybe read from `LISTEN_ADDRESS(4|6)` and bind to `[::]` if not set? I think binding to IPv6 addresses may have edge cases since IP allocation isn’t as straightforward as dhcpv4.

Will probably recommend folks to containerize and use associated fw rules to isolate if bound ipv6 addresses turns out to be problematic

Reply:
It is a bit of boilerplate but we could just bind to all interfaces that we find? I am starting to think it might be better if we use what is there (any combination of IPv4/IPv6 that is available). It makes the code a bit more complicated because we have to make some runtime decisions then but it is much cleaner and allows us to run in more environments.

Reply:
So I thought that we could be clever here and just always run on dual-stack but it is actually a bit more complicated than I anticipated. I've recorded an issue here https://github.com/firezone/product/issues/595 and I am thinking of stripping the IPv6 support again and do it properly at a later point.